kaspresso/src/main/kotlin/com/kaspersky/kaspresso/runner/listener/KaspressoRunNotifier.kt

```diff
@@ -0,0 +1,70 @@
+package com.kaspersky.kaspresso.runner.listener
+
+import android.app.Instrumentation
+import org.junit.runner.Description
+import org.junit.runner.Result
+import org.junit.runner.notification.Failure
+import java.lang.IllegalStateException
+
+interface KaspressoRunNotifier : KaspressoRunListener {
+    val listeners: List<KaspressoRunListener>
+    fun addListener(listener: KaspressoRunListener)
+}
+
+inline fun <reified L : KaspressoRunListener> KaspressoRunNotifier.addUniqueListener(createListener: () -> L) {
+    if (listeners.filterIsInstance<L>().isEmpty()) {
+        addListener(createListener())
+    }
+}
+
+inline fun <reified L : KaspressoRunListener> KaspressoRunNotifier.getUniqueListener(): L {
+    return listeners.filterIsInstance<L>().takeIf { it.size == 1 }?.first()
+        ?: throw IllegalStateException("Unique listener is not single")
+}
+
+fun Instrumentation.asKaspressoRunNotifier(): KaspressoRunNotifier =
+    requireNotNull(this as? KaspressoRunNotifier) {
+        "com.kaspersky.kaspresso.runner.KaspressoRunner is not set"
+    }
+
+internal class KaspressoRunNotifierImpl : KaspressoRunNotifier {
+    private val cache = KaspressoLateRunListener.Cache()
+    override val listeners: MutableList<KaspressoRunListener> = arrayListOf()
+
+    override fun addListener(listener: KaspressoRunListener) {
+        if (listener is KaspressoLateRunListener) {
+            listener.lateInit(cache)
```

Comment:
add some info in comments why do we need this method invocation. (Because some listeners are added after tests started and they left call back invocations)

Reply:
Allure listener and allure runner